src/oph/ehoks/hoks/hoks.clj

```diff
@@ -345,6 +375,25 @@
                              h))))
     hoks-db))
 
+(defn check-and-save-hoks!
+  "Tekee uuden HOKSin tarkistukset ja tallentaa sen, jos kaikki on OK."
+  [hoks]
+  (let [opiskeluoikeudet
+        (k/fetch-opiskeluoikeudet-by-oppija-id (:oppija-oid hoks))]
+    (when-not
+     (oppijaindex/oppija-opiskeluoikeus-match?
+       opiskeluoikeudet (:opiskeluoikeus-oid hoks))
+      (throw (ex-info "Opiskeluoikeus does not match any held by oppija"
+                      {:error :disallowed-update})))
+    (when-not
+     (oppijaindex/opiskeluoikeus-still-active? hoks opiskeluoikeudet)
+      (throw (ex-info (format "Opiskeluoikeus %s is no longer active"
+                              (:opiskeluoikeus-oid hoks))
+                      {:error :disallowed-update})))
+    (validate-tuva-hoks-type hoks)
+    (validate-hoks-osaamisen-hankkimistavat-with-opiskeluoikeus! hoks)
+    (save-hoks! hoks)))
```

Comment:
Tässä ekana haetaan oppijan kaikki opiskeluoikeudet, sitten katsotaan, mikä opiskeluoikeuksista mätsää hoksista löytyvään (vai mätsääkö mikään). Pystyisikö `opiskeluoikeudet` tietueesta filtteröimään tuossa ihan alussa hoksin opiskeluoikeus-oid:ta vastaavan opiskeluoikeuden, jolloin opiskeluoikeutta ei tarvitsisi myöhemmin hakea uudestaan Koskesta funktiossa `validate-tuva-hoks-type` eikä eHOKSin indeksistä funktiossa `validate-hoks-osaamisen-hankkimistavat-with-opiskeluoikeus!`?

`check-and-save-hoks!` taidetaan myös kutsua tuolla `post-hoks` funktiossa, jossa kutsutaan `op/add-hoks-dependents-in-index!`, jossa **myöskin haetaan opiskeluoikeus Koskesta**. Vaikuttaa siis, että aika monta turhaa kutsua tehdään Koskeen. :smile: Tuon opiskeluoikeus tiedon hakemisen voisi yrittää saada sijoittumaan funktioiden call stackissä mahdollisimman matalalle, jolloin opiskeluoikeustieto voidaan hakea kerran ja passata parametrina sitä tietoa tarvitseville funktioille, jonka seurauksena aika monesta funktiosta tulee myös käytännössä sivuvaikutukseton.

Sama juttu koskee `check-hoks-for-update!` funktiota, jossa opiskeluoikeus haetaan Koskesta `oppijaindex/opiskeluoikeus-still-active?`- ja `validate-tuva-hoks-type`-kutsuissa ja vielä kerran eHOKSin indeksistä `validate-hoks-osaamisen-hankkimistavat-with-opiskeluoikeus!` kutsussa.

Jos tuota opiskeluoikeutta tarvitaan tosi monessa checkissä, joiden "epäonnistuminen" käytännössä johtaisi siihen että koko request failaa, niin voi olla käytännössä ihan sama tehdä tuo opiskeluoikeuden haku niin varhaisessa vaiheessa kuin suinkin on mahdollista?

Reply:
> Tässä ekana haetaan oppijan kaikki opiskeluoikeudet, sitten katsotaan, mikä opiskeluoikeuksista mätsää hoksista löytyvään (vai mätsääkö mikään). Pystyisikö `opiskeluoikeudet` tietueesta filtteröimään tuossa ihan alussa hoksin opiskeluoikeus-oid:ta vastaavan opiskeluoikeuden, jolloin opiskeluoikeutta ei tarvitsisi myöhemmin hakea uudestaan Koskesta funktiossa `validate-tuva-hoks-type` eikä eHOKSin indeksistä funktiossa `validate-hoks-osaamisen-hankkimistavat-with-opiskeluoikeus!`?

Yritin rajoittaa näiden refaktorointia, koska tässä muutenkin refaktoroitiin jo aika paljon (olen siis tosissani yrittänyt miettiä, mikä on _vähin_ refaktorointi, jolla saan uuden työpaikkajaksojen tarkistuksen vain kahteen paikkaan kuuden sijaan).  Harkitsin itse myös opiskeluoikeuksien keräilyä yhdessä kohdassa.  Oikeastihan on siis tietysti täysin älytöntä, että keräillään oppijan kaikki opiskeluoikeudet ja etsitään niistä HOKSia vastaava.  Tämä tehdään vielä kahdesti, sekä oppija-opiskeluoikeus-match?:ssa että opiskeluoikeus-still-active?:ssa.  Ihan hyvin voisi vain hakea HOKSin opiskeluoikeuden ja katsoa, että se on voimassa ja HOKSin oppijan opiskeluoikeus.

Alkuperäinen syy lienee se, että hankintakoulutusten lisääminen indeksiin vaatii joka tapauksessa koko oppijan opiskeluoikeuksien listan.  Niinpä "optimointina" haetaan oppijan kaikki opiskeluoikeudet kerralla ja myös tarkistukset tehdään oo-listan, ei HOKSin oo:n perusteella.

> `check-and-save-hoks!` taidetaan myös kutsua tuolla `post-hoks` funktiossa, jossa kutsutaan `op/add-hoks-dependents-in-index!`, jossa **myöskin haetaan opiskeluoikeus Koskesta**. Vaikuttaa siis, että aika monta turhaa kutsua tehdään Koskeen. 😄 Tuon opiskeluoikeus tiedon hakemisen voisi yrittää saada sijoittumaan funktioiden call stackissä mahdollisimman matalalle, jolloin opiskeluoikeustieto voidaan hakea kerran ja passata parametrina sitä tietoa tarvitseville funktioille, jonka seurauksena aika monesta funktiosta tulee myös käytännössä sivuvaikutukseton.

Kaikki tää lienee oikeasti ihan samantekevää, koska `get-oppijat-opiskeluoikeudet` ja `get-opiskeluoikeus-info-raw` käyttää fifo-ttl-cachea ja niinpä, niin kauan kuin me ei saada eliminoitua jompaakumpaa kutsua kokonaan, molemmat tehdään ja kumpikin vain kerran.  Niinpä kyse on pitkälti estetiikasta.  Mulle on esteettistä, että oppijaindeksin päivitykset ja HOKSin validaatiot tehdään eri funktioissa ja niinpä, jos halutaan threadata niihin sisään samoja opiskeluoikeuksien tietoja, ne täytyy hakea jossain funktiossa joka kutsuu molempia ("matalalla call stackissa" tarkoittaa varmaan tätä?) ja jonka vastuulle ei kuulu tietää, mitä kaikkea tietoa validointeihin tai indeksöinteihin tarvitaan.  Mut oikeasti tällaisissa asioissa on puolensa ja puolensa eikä tietenkään kaikkien maku sano samaa.  Mutta jos haluat refaktoroida tän niin, että opiskeluoikeus (tai -oikeudet) haetaan vain kerran, veikkaan etten osaa tehdä sitä sellaisella tavalla josta me molemmat tykätään, joten jättäisin sen mieluusti sulle enkä aio urputtaa siitä jos siinä on mun mielestä huono vastuunjako funktioiden välillä :)  Ainoo tapa, jonka mä keksin, millä tästä voisi saada edes jotenkuten kauniin, on systeemi jossa requestiin (tai johonkin muuhun catch-all kontekstimuuttujaan) lisätään kaikenlaista tietoa (delay)na, siten ettei sitä lasketa ellei sitä oikeasti joku kaipaa.  Monet sit taas inhoaa tällaisia kontekstimuuttujia, joissa on ihan mitä sattuu tietoa.

Jos sen sijaan ei välitetä muista koski-kutsuista, vaan ollaan kiinnostuneita vain näistä validaatioista, niin niille tosiaan on nyt mahdollista (kun ne on vihdoin samassa funktiossa) threadata tästä `check-and-save-hoks!`:sta opiskeluoikeus.  Sen tekisin mieluiten erillisessä PR:ssä.

> Sama juttu koskee `check-hoks-for-update!` funktiota, jossa opiskeluoikeus haetaan Koskesta `oppijaindex/opiskeluoikeus-still-active?`- ja `validate-tuva-hoks-type`-kutsuissa ja vielä kerran eHOKSin indeksistä `validate-hoks-osaamisen-hankkimistavat-with-opiskeluoikeus!` kutsussa.

Jep, en usko että on sattuma että Kosken clientissa on se ttl-cache :)

> Jos tuota opiskeluoikeutta tarvitaan tosi monessa checkissä, joiden "epäonnistuminen" käytännössä johtaisi siihen että koko request failaa, niin voi olla käytännössä ihan sama tehdä tuo opiskeluoikeuden haku niin varhaisessa vaiheessa kuin suinkin on mahdollista?

Tästä ei huolta, se failaa jo siinä vaiheessa kun ollaan `add-hoks-dependents-in-index!`:ssa `insert-opiskeluoikeus` kutsuu `get-opiskeluoikeus-info`:a

Reply:
Tää on muuten monimutkainen aihe jossa saattaa tulla väärinkäsityksiä, pitäiskö ottaa videopuhelu aiheesta?

Reply:
tästä pitää tiketöidä idealistalle mahdollisia arkkitehtuurimuutoksia...